lms/static/scripts/frontend_apps/utils/use-warn-on-page-unload.ts

```diff
@@ -0,0 +1,29 @@
+import { useEffect } from 'preact/hooks';
+
+/**
+ * Registers an event listener to window's 'beforeunload' if `hasUnsavedData` is true.
+ * It also unregisters the event if `hasUnsavedData` is false or the component is unmounted.
+ *
+ * This event listener makes the browser warn the user about potential unsaved changes,
+ * and gives the user the opportunity to cancel the page unload if desired.
+ *
+ * @link https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event
+ */
+export function useWarnOnPageUnload(hasUnsavedData: boolean) {
+  useEffect(() => {
+    const listener = (e: BeforeUnloadEvent) => {
+      e.preventDefault();
+      e.returnValue = '';
+    };
+    const removeListener = () =>
+      window.removeEventListener('beforeunload', listener);
+
+    if (hasUnsavedData) {
```

Comment:
If `hasUnsavedData` is false, the effect should just return early near the top. There is no need to remove the listener because the listener from the current effect won't be registered. In a sequence of updates where `hasUnsavedData` transitions from false => true => false, the sequence of events should be:

1. First effect runs with `hasUnsavedData == false`, effect should be a no-op
2. Cleanup of first effect is run. It should do nothing.
3. Second effect runs with `hasUnsavedData == true`, a listener will be registered
4. Cleanup of second effect runs. It should remove the listener.
5. Third effect runs with `hasUnsavedData == false`, effect should be a no-op